torch/csrc/autograd/python_saved_variable_hooks.cpp

```diff
@@ -9,19 +11,35 @@ namespace torch { namespace autograd {
     unpack_hook_(unpack_hook.release().ptr()) {}
 
   void PySavedVariableHooks::call_pack_hook(at::Tensor &tensor) {
-    TORCH_CHECK_NOT_IMPLEMENTED(false, "Hooks are not implemented yet");
+    py::gil_scoped_acquire acquire;
+    auto pack_hook = py::reinterpret_borrow<py::function>(pack_hook_); // borrows the reference to the hook
+    auto wrapped = THPVariable_Wrap(tensor); // here, we own the reference
+    py::object obj = py::reinterpret_steal<py::object>(wrapped); // obj steals the reference
+    py::object packed = pack_hook(obj); // packed is a new object with a reference
+    data_ = packed.release().ptr(); // steals the reference
+    // pack_hook, obj are decrefed on exit, but we will manually decref data_ when the saved variable is released
+    // pack_hook_ is left as is
   }
 
   at::Tensor PySavedVariableHooks::call_unpack_hook() {
-    TORCH_CHECK_NOT_IMPLEMENTED(false, "Hooks are not implemented yet");
+    py::gil_scoped_acquire acquire;
+    auto unpack_hook = py::reinterpret_borrow<py::function>(unpack_hook_);
+    py::object obj = py::cast<py::object>(data_); // copies the content of data_
```

Comment:
This doesn't actually perform a copy does it? It does get a new reference though?

Reply:
Yes, sorry my bad, it's only a copy in the sense that we get a new reference.